core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -589,10 +600,10 @@ public void process(@Nullable MappingMetaData mappingMd, boolean allowIdGenerati
         }
 
         // generate id if not already provided and id generation is allowed
-        if (allowIdGeneration) {
-            if (id == null) {
-                id(UUIDs.base64UUID());
-            }
+        if (allowIdGeneration && id == null) {
+            assert autoGeneratedTimestamp == -1;
+            autoGeneratedTimestamp = System.currentTimeMillis();
```

Comment:
The "problematic" edge case is where someone has their time way in the future. Note both cases are "safe" in the sense that no duplicates will occur. Have the time in the future will disable the optimization for indexing request coming from other nodes. Having it in the past will cause problems on for this specific one. To deal with the first one we added the current timestamp used by the engine to the segments stats. See [here](https://github.com/elastic/elasticsearch/issues/19813#issuecomment-243079678).